### ISSUE TITLE
Designing the number of attempts feature.

### ISSUE THREAD

**Issue by aharjula** [maths/moodle-qtype_stack#182]
As some of us know there has been some interest in giving the question author a variable that would tell how many attempts the student has done. Basically, that number would be handy if one wants to start giving more and more feedback if the student keeps making errors repeatedly, the most obvious use case is naturally hints.

As just having the number of previous attempts is basically impossible in the question model of Moodle this feature requires some extensions before we can even start counting attempts. So this is a feature that needs the state-storage extensions of the stateful variant of STACK. That variant has now stabilised quite a bit and has reached a point where the state storage mechanism will be frozen now would be a good time to start designing the basic features provided automatically (but only when requested) by using state. And the obvious one to start with is the number of attempts...

I have gotten quite far with this already but now would be the time to do a sanity check and add those obvious features I have missed, to start the discussion here is a screenshot of the first ever stateful STACK question with the addition of full debug output of the state variables it has access to and the attempt counts printed out after that. Those not familiar with the question should know that it asks for roots of a polynomial one at a time and modifies the question text, feedbacks and PRTs as the student gives correct roots, naturally the conditional rendering features of the abacus branch are in heavy use here.

![attempts](https://cloud.githubusercontent.com/assets/980957/12948358/46ebf4e4-d00a-11e5-9ea8-e4059f13dc25.png)

Some explanations about the variables:
- The first variables listed are the `instance`-variables that are tied to this question instance and are used to transfer the state of the question between attempts. In this case we store the set of the roots the student has already found as well as the roots in the order be have processed them. The attempt counter could be handled manually using the same logic just increment a number every time we get an attempt. In this case we handle it by automatically storing the inputs to the trees in the `[prt1]` and `[prt2]` variables, `[prt3]` has not appeared yet as there have been no input to it.
- Then come the automatically populated `structure` variables that provide various maxima functions a way to deal with named variables that happen to have identifiers in them like `previous-input-ans1`.
- Finally, come the `prt=?` scoped variables that are generated on the maxima side from those `structure` variables and from the input history provided in the `[prt1]` like variables. Currently I have provided the following variables:
  - `required-inputs` is something that a coder might need but probably does not interest anyone else.
  - `history` is a list of all the valid inputs this tree has received in maxima form and in order, to be added on this list the input vector (there may be multiple inputs tied to the tree) in its raw input string form must differ from the previous attempts raw input string but as maxima may simplify things the actual values in this list may not differ that much. `[prt1]` may explain the situation better.
  - `count-attempts` is obviously just the length of `history`, note that history contains the current attempt as its last element.
  - `previous-input` provides a way of accessing the previous input vector without having to think about the indices and reversals one would have to do to the `history` variable.
  - `previous-input-?` provides direct access to the values of the previous attempts inputs one at a time so that you do not need to find out in which order the inputs are in that input vector. Anyhow that ordering is provided for those that need it in code.
- Some values are `und` undefined, for example we can't have a value for previous attempt if there has not been such an attempt. So code referencing such a thing will need to be able to deal with that. Naturally, this might cause problems if the input actually was `und`.

Some open questions:
1. Does anyone need a count of distinct attempts i.e. not just different by string value but also algebraically? If so we naturally would only compare to the previous input and so on to not break the ordering, but would you then need access to the previous distinct input?
2. Does everyone understand why we need to do this counting separately for each tree? Basically, does someone have some grand idea for a counter that would count the whole questions attempts? Would such even be useful? Keeping track of such an value in STACK question model might prove impossible so lets hope someone explains this in such a way that we do not hear that request every day.
3. Can anyone imagine a situation where we would need to have shorthand access to the penultimate inputs? Or earlier? What is the term for one before penultimate if we follow this path? In my opinion if someone wants to access those inputs they may just as well use the lists.
4. Did we have some own internal symbol for `und`? So that we could leave that as an unambiguous input value?

The code for this will probably appear in my repository around the end of the month and it will mark the transfer from alpha to beta so that people may actually start testing it with bit more confidence. So now would be a good time to add those features and change stupid things before beta-users learn to live with that stupid things.

**Comment by timhunt** [comment]
1. This is something Chris and I talked about way back when we started work on STACK 3 (putting STACK into Moodle). There are clearly uses for this, but there are also issues.

2. The main difficulty is that it breaks some assumptions that help simplify how the code works: For example, the result of evaluating a PRT is a pure function (no side-effects) depending only on (question variables, current state of all the inputs). The only way history affects things is in adding up the penalties each time a PRT is evaluated.

3. But, therefore, we do already have a clear notion of 'number of tries', but it is (potentially) different for each PRT. See this unit test for an example of the general case: https://github.com/maths/moodle-qtype_stack/blob/7ecf85b5c6c78f935fae09dd99dfeea5ee977fd4/tests/walkthrough_adaptive_test.php#L1049

4. However, it is probably only a small step from there to making it a pure function of (question variables, current state of all the inputs, try number for this PRT). However, you need to get 'try number' from the question behaviour (interactive or adaptivemultipart).

5. If you look in either of those behaviours, you will see that the system already counts tries. However, it does not currently pass the try number on to the question. (The assumption about what can affect the grading goes beyond STACK.)

6. If we do this, how are we going to enable people to write question tests for the new greater complexity?

**Comment by aharjula** [comment]
Tim.

1. Oh I know this is one of the things many have talked about for quite some time. I for one have thought about the state-expansion since early STACK 2 time and as I start to get that working I thought maybe I could give this feature to those that want it through it. Naturally, there are always issues when we change stuff and for that reason I have ensured that normal old STACK questions will work just the same in the state expansion as they did before and you start to see those issues only if you mix state into your questions. The current open issues I have are as follow:
 
 - Backup/restore of question attempts is a bit difficult if the state is not stored in step-data. But that is something that can be solved by simply burning insane amounts of CPU cycles. Currently we do not support backup/restore at all.
 - The attempts are marked "not complete" due to input fields not being filled in questions that do not always show all the input fields, which is inconvenient but again something that can be solved by bringing question-text awareness to the functions that tell if something is complete. The same problem appears with the check-button when the question-text no longer has any inputs visible.
 - Clean-up of state stored for attempts that have been deleted needs a hook to the cron-job or if the world was better we would just add ``ON DELETE CASCADE`` to the database.

2. That is a beautiful simplification that some people need but some other people are willing to go further and deal with the costs of dropping that. Naturally, dealing with the complexity arising from adding state is something that the question author must accept if they choose to add state to their questions. But I for one do not want to say that people should not have some toys just because some people might hurt themselves with those toys. Actually, in reality I am the one that wants to add dangerous toys to the toy-box to weed out people that do not understand the situation. But I am not against warning labels in this case.

3. Yes I am aware that the behaviour layer does various things but as the Moodle question model blocks our access to that layer and seems to want to force the pure function assumption of your point 2 we have no way of accessing that information. Unless we make our own behaviour that talks to our question in specific ways but we have talked about that previously have not...

4. Actually you do not need to get that from the behaviour, you can just write counters out as hidden inputs and handle all the incrementation on the question side (well we could if we could assume that the question only gets asked to grade stuff if the stuff has changed, but it is not so even with behaviour that have functions which seem to imply such), basically the state system now has a single counter that keeps track of the step in progress and uses that to retrieve the corresponding state from the database. But the raw number is not enough as we care about the actual input, is it different than the previous one, mathematically or just in the string comparison sense? And what if we want to see the previous attempts value in the question logic?

5. Again we would want to push that value all the way to the Maxima and maybe even so that it would affect the question text. The problem is that the processing order that the Moodle question model has does not give us many places where that could happen and there are various causality issues in pretty much everyone of the places where that might be possible. Just to get my own state system working I had to reorder things in the rendering side so that things are present in the correct order, luckily the ``render.php`` is still part of the question type.

6. Simply, we do not. If a person wants to be a coder that handles state variables in questions he must also be able to prove the working of that question in other ways, especially in ways that are not just about testing edge cases. The world is full of testing methodologies that do not need the system to provide tools for testing. If we really wanted to provide unit testing we would need to allow definition of input sequences, which might also help in testing the penalty accumulation in normal STACK questions, but why would we need that. In my opinion unit testing is not truly testing at all it is just a hope that things that break break also the special cases we test but hey if you simplify your world enough you can assume that unit testing is enough.


Just for comedic comment, if this pure function concept is so important what would you say if I managed to magically bring multi threading into the picture... seriously I have no idea for what I would use that but surely I could map the PRT handling as concurrent processes although that would break the state transition system I just wrote.

**Comment by aharjula** [comment]
Hope someone sees this issue and can give some views about this from the user perspective. Basically, what do you need...

Anyway the code for that and the new button input type are now in my alpha2 repository.

**Comment by sangwinc** [comment]
Matti, thank you for prodding me again about this.  I've been tied up and distracted with other things, so I'm sorry about the delay in replying.

My only concern is that things suddenly get complicated!  STACK is already a beast, and question authors struggle.  I think we should aim for the features we have really strong use-cases for in the first pass.  Toys are fun, but I've got a job to do and I want to support colleagues.  Did I mention the French restaurant vs British pub before?   Let's have a short high quality menu we can support well, with really useful reliable tools.  What you have done here is great, and I've got some interesting ideas for questions of my own.  The most difficult part is to find a powerful and simple model.
 
* Your questions.

1. Does anyone need a count of distinct attempts?

Not at the moment, other than in the code (which already exists) which calculates the penalty.  If we introduce such a count I think they should always be identical.

2. Does everyone understand why we need to do this counting separately for each tree? 

YES!  I can't see how we could do this other way.  

3. Can anyone imagine a situation where we would need to have shorthand access to the penultimate inputs?

Not for me at this stage, and I think users who are sophisticated enough to use state should be able to use maxima like `second(reverse(list))` to get the penultimate.  We should document such knowledge, but try to avoid duplication.  We should also keep the number of separate state variables to a minimum.

* Comments

I can see a need for "previous-input" and "previous-input-ans1".  Is this just convenience, but it is useful.

Yes, we need to agree a "null" value of some kind.  "und" is already Maxima's undefined: "und represents an undefined result.", so this is a good choice.  Of course, this might be an answer.  So, do we need a "stackund" to be clear here?

Testing of questions.  Matti, I'm increasingly in favour of having more testing, which is more comprehensive.  E.g. I spent my Saturday adding in more testing to the answer tests (https://github.com/maths/moodle-qtype_stack/commit/fb3af059e8eea8603309bb9c87631ea874a78f16).  I should have done this *years* ago.  I found a bug! ;-)  I also think we should harden up the existing question testing to record the route through the tree, not just the final note.

As you say, unit testing has its limits, of course.  For all its flaws, we must have some kind of automatic testing.   There are a number of reasons for testing.  (1) getting it to work, (2) communicating to a colleague what the question does (and special cases are really good here), (3) seeing if something changed.  I urge you to reconsider your response (6) to Tim here. Please please please design in some testing mechanism.

* My questions.

(A) If prt1 has required input ans1 and ans2, what will the "history" variables look like?  List of lists?

(B) There is no question here of passing state between PRTs is there!  Can we absoluetly draw the line there and say "no"!

(C) At a minimal and abstract level, are you planning to make the "history" available to the PRT?  Everything else can be recreated from that?

I.e. count-attempts = length(history)
     last-attempt = first(reverse(history))
     
Is that really the case?  If so, here is perhaps a radical suggestion.  The state code (in PHP) is confined *only* to the history for now.  On the Maxima side we build a PRT data structure from the history which gives a teacher access to these variables directly.

Tim's response (2) Can we keep the pure function model attempt?  I'd be really happy if we change STACK from

`  attempt_outcome = prt(question variables, current state of all the inputs)`  
to 
`  attempt_outcome = prt(question variables, history of all the valid inputs)`

That is not such a big leap.  Did you specify a use case which needs more than this?  I can't remember, sorry Matti!  Can you remind me of this?

By doing this we keep the number of state variables to an absolute minimum.  I know this potentially means "undoing" a lot of good work, but that work has helped us understand the problem.  I had to do a lot of this myself when I realised we could decouple the inputs and prts.  This simplified everything, and I threw away loads of "different question types" that hard-wired inputs to prts in various ways.  I threw away a lot of code recently when I realised the best way to randomly generate MCQs was in maxima....

So, this is a very long message.  What you propose is radical, and a lot of work.  Thanks Matti, this looks very valuable.  My suggestions would be to try to keep it simple, even at the cost of not covering all our intended cases in round one.

**Comment by aharjula** [comment]
The issue of things getting complicated is indeed a valid point and for that reason I have tried to keep this feature away from the base level question authors way (and I think I have succeeded in that). So they do not need to touch it and it does not affect their lives unless they touch it.

"stackund" would probably solve the edge case of "und" as input. If we go that way it would be nice to use the same one everywhere where this kind of a situation may appear.

As to testing, here is a point where I have hard time of understand the balance between complexity for the question author compared to the gains. To truly test a state variable using question one would need to:

1. Define the initial state, if there is global variables present or if we are continuing from the middle of a question attempt.
2. Declare inputs to be given and the results of the PRTs activated by those inputs as well as the state after those activations.
3. And repeat that for multiple inputs to the same active question. Or skip the automatic transfer of end state of the previous input processing to the entry state of the next. And only allow tests that define the entry and exit states with inputs.
4. There truly is no point in testing PRTs independently when you have state mixed in the question, you really want to test the whole question and series of inputs.
5. You may also want to test the contents of the question text as that will be something that has a bigger role.

Basically testing would require such a large form and understanding of the breaking of the PRT-level abstraction when we work with whole questions that I do not see the point when the question could just be designed correctly.

(A) Indeed list of lists, it is that already in the example but the list only has one element in it. The ordering of that list is the same as the declaration order of the inputs required by that PRT. But as unwrapping the values from that is hard and we do not want to see magical numbers being used to do that I wrote the code to build the convenience variables like "previous-input-ans1" and those are generated on [the maxima side](https://github.com/aharjula/moodle-qtype_stack/blob/STACK_4.0_ALPHA2/stack/maxima/stackstate.mac#L125-L174) from the raw history list using the question-structure information, they do not get stored or passed between maxima and STACK. The comment about keeping the number of separate state variables low does not matter here as those variables are created inside maxima-functions and do not even appear in the CAS-response text unless we actually reference them. Also the history state-variables are only initialised for the PRTs if someone references them in the code so we are saving CAS, CAS-cache, and CAS-connection resources as much as we can.


And here we start to get to the problem of just having the number of attempts or having the state-variables from my vision. As you say you do not need state-variables to have the history if you generate that history in PHP and push it to the session where you need it. The state variable infrastructure just happens to provide a way of doing that but it does so much more. You are naturally free to implement the number of attempts any way you want, but as I have a way that basically provides a way of testing it using the state-variable system I thought that it would be sensible to think about what that number of attempts thing really consists of, not necessary how we are implementing it.

(B) No... You can write stuff to "instance"/"global"-variables inside the PRT-code and then read it and anything else stored in the state-system (including other PRTs histories, which might not yet include the currently processing steps inputs) in another PRT. As the state transitions can only happen in the PRT feedback variables we need to write stuff there and as the state can affect the processing in that PRT we also need to read it there. Actually that polynomial-question example already does that, the first PRT stores the found root and the second PRT checks that you are not giving the same one to it. Processing order has been defined so that the PRTs are processed in the order they exist in the question editor. 

(C) Yes the history and all state is available everywhere where we have cas-sessions. But here comes the place where I have made the technical things more complex because I do not want to make the question authoring even more complex. If we were to have just 
```attempt_outcome = prt(question variables, history of all the valid inputs)``` we could not have global state as there would not be a way of getting state out of the system and we would require all the state values to be written as functions of the history in basically functional-style code which only references the history. Instead I have made it so that you may create your own variables how ever you want and we pass their values as is between steps without you having to wonder how to write the function that maps the history to whatever you need whenever you need it and getting that evaluated so that it would be usable in the question-text is a problem, basically there would be insane amount of code inside the question-text instead of the large amount we have with this. In essence I saw that while some subset of this could be done as pure functional programming there many quite important features that need the simpler more procedual style of programming and for this reason we are not purely functions of the raw history in the PRT scale in the whole question level one can still claim that but only if we have no global state.

My model is:
```attempt_outcome = question(question variables, current state of all the inputs, selected values from previous attempts processing)``` and PRTs do some logical processing inside that in such a way that we can still consider PRTs as parts or a multi part question in the traditional way, but the moment they depend from state variables they are no longer independent actors like in the traditional STACK models PRTs.

**Comment by sangwinc** [comment]
Matti,

We must be able to come up with a way of testing the questions efficiently!  How else will anyone be able to understand what is going on?

The HTTP request to submit gives us a very definite "event" (even with ajax validation), and we can use this list of attempts to build up a table with one row for each event, and one column for each input.  The testing form could replicate this, with each input column essentially being the history+empty+invalid inputs.

For each PRT we have such a table supplemented by columns giving outputs such as the score, note and any state.  

Even if we don't automatically test everything (we only currently test the final note...) we must have some mechanism for creating a history and running a question over that history.  Otherwise nobody will ever be able to write or test questions.  

When I think back over all the projects I've worked on, I have to reflect on what some were picked up and used, (e.g. dragmath) and others not so much.  One of the reasons Geogebra is so popular is the low barrier to entry.  Moodle is modular making it easy to contribute.  My concern here is that we risk introducing very complicated features that are opaque, i.e. users cannot see what is going on.  A good testing framework (semi-automatic is fine) will really help users adopt the system,

What do you think?

**Comment by aharjula** [comment]
Chris,

That is the way the testing would have to be done and if this is the thing that is needed to stop another fork it will be done, but don't say that I did not warn you about the testing system being scary looking with state using questions, remember that those questions typically have multiple scenes with multiple inputs each with PRTs for each scene transition as well as PRTs inside the scenes, the number of columns in that table will be immense and writing tests using it will again only cover just the transitions and PRT tests the test writers bothers to write. I believe that to do any reasonable testing for a basic state using question will probably mean filling a 30x15 matrix, typically much wider when working with the typical business or statistics question we have heard visions of, and the height obviously will grow with the level of feedback and I believe these questions will naturally give more feedback than the typical STACK questions, even the feedback may have a storyline here. If the question has even one random parameter and I were a question author I would not even think about filling that matrix I would rather prove the transitions on paper for all the random parameters of the question and hope that the answer tests provided by STACK keep their form.

Naturally, we could bring down dimension of that test matrix if we were to allow writing tests with some code that defines inputs and state and tests the outcome for just specific features instead of the whole outcome, i.e. "does the question text hide these inputs after this step?", "did state variable X change?",  and so on, bit like those walkthrough tests. But that would again require new syntax parsing and grammar design, and considering how far the previous grammar idea of {#...#} has gotten that would probably never happen.

If you want to expand the current testing system to cover this you would simply add a new field where the test writer could write the current state and if the test writer knows what they are doing they can even write the history vector through that, but it would then only test a single PRT and as the state system passes the state between the PRTs we need to test all the PRTs at the same time.

Obviously the state system is complicated (how could it not be) and building questions using it requires a bit more understanding, but as it brings nothing new to the editor (well the variable definitions, but that could be hidden by default) the normal user will not need to even know about it. The reality is that if some author then chooses to use it there will always be a barrier to entry as the author needs a bit deeper understanding of the question model and maybe even know a bit about programming. Sure some authors will fail to climb over the barrier but they will still have the other tools of STACK to use, but some will get over the barrier and some of them will make questions that will make the world a better place and those authors are the authors that truly understand how to implement their visions and we want to hear their ideas on developing future question models and features (not necessarily to implement but maybe to just acknowledge or to redirect to other solutions), otherwise they will build their own systems from scratch and maybe we will never hear about them or maybe we will then see things that are basically STACK with just some small differences that again causes potential users to hesitate if they should use these systems when they can't choose between them.

**Comment by sangwinc** [comment]
Thanks Matti,

I think the key to this is to develop *some* way. If, as you describe, the whole things gets too complex for a single table then I can imaging some hybrid between unit testing and BEHAT.

E.g. The question author might just indicate the history as a Maxima list of potential inputs, so no complex forms.  I agree: we can expect such advanced users to create a list of things.  We then have a page with a "next button" and the author steps through the question, seeing the question stem evolve (filled in with the inputs), with the feedback etc. and a separate section listing details of the internal state below.

If we have a "confirm behaviour" button, we can then test each random version has the same basic behaviour for similar inputs, defined in terms of random variables.  This way, a teacher can be confident that correct answers are marked correct, etc.  This would be much better (in my mind) than nothing.  It would help us develop use-case examples to explain to other people what is going on.

I reinvented unit testing myself out of frustration!  Unusually, I decided to expose these tests to help the normal user (question author, not developer) see what the input-output behaviour of the answer tests does.  I think this was very sensible.  I'd really like something similar here.
